kafka-rest/src/test/java/io/confluent/kafkarest/integration/CustomLogIntegrationTest.java

```diff
@@ -0,0 +1,515 @@
+                i -> {
+                  if (path.contains("/records")) {
+                    // Produce API call
+                    ByteString key = ByteString.copyFromUtf8("foo");
+                    ByteString value = ByteString.copyFromUtf8("bar");
+                    ProduceRequest request =
+                        ProduceRequest.builder()
+                            .setKey(
+                                ProduceRequestData.builder()
+                                    .setFormat(EmbeddedFormat.BINARY)
+                                    .setData(BinaryNode.valueOf(key.toByteArray()))
+                                    .build())
+                            .setValue(
+                                ProduceRequestData.builder()
+                                    .setFormat(EmbeddedFormat.BINARY)
+                                    .setData(BinaryNode.valueOf(value.toByteArray()))
+                                    .build())
+                            .setOriginalSize(0L)
+                            .build();
+                    return executor.schedule(
+                        () ->
+                            request(path)
+                                .accept(MediaType.APPLICATION_JSON)
+                                .post(Entity.entity(request, MediaType.APPLICATION_JSON)),
+                        /* delay= */ i * requestInterval.toMillis(),
+                        TimeUnit.MILLISECONDS);
+                  } else {
+                    return executor.schedule(
+                        () -> request(path).accept(MediaType.APPLICATION_JSON).get(),
+                        /* delay= */ i * requestInterval.toMillis(),
+                        TimeUnit.MILLISECONDS);
+                  }
+                })
+            .collect(Collectors.toList()).stream()
+            .map(
+                future -> {
+                  try {
+                    return future.get();
+                  } catch (InterruptedException | ExecutionException e) {
+                    throw new RuntimeException(e);
+                  }
+                })
+            .collect(Collectors.toList());
+
+    for (Response response : responses) {
+      assertTrue(
+          !response.getHeaders().containsKey(CustomLogRequestAttributes.REST_ERROR_CODE),
+          "Unexpected header in headers " + response.getHeaders());
+      int status = response.getStatus();
+      if (status != 200 && status != 429) {
+        fail(
+            String.format(
+                "Expected HTTP 200 or HTTP 429, but got HTTP %d instead: %s",
+                status, response.readEntity(String.class)));
+      }
+    }
+  }
+
+  private void verifyLog(
+      int expectedNumOfEntries,
+      int errorCodeInLog,
+      int expectedRateLimitLogs,
+      boolean isProduce,
+      boolean isCustomLoggingDisabled) {
+    // Sleep so all request-logs are logged before reading and validating them here.
+    try {
+      TimeUnit.SECONDS.sleep(2);
```

Comment:
this could potentially be the source of flakiness, is there another source to check that all requests have finished?

Reply:
That's fine